File: backend/src/main/java/com/example/backend/model/Transaction.java
package com.example.backend.model;

import com.example.backend.common.PaymentMethod;
import com.example.backend.common.TransactionStatus;
import com.example.backend.common.TransactionType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "transactions")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "transaction_id")
    private UUID id;

    // 1:1 relationship with Booking
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "booking_id", nullable = false)
    private Booking booking;

    @Column(name = "amount", nullable = false)
    private Double amount;

    @Column(name = "currency", nullable = false, length = 10)
    private String currency;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", nullable = false, length = 50)
    private PaymentMethod paymentMethod;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private TransactionStatus status;

    @Column(name = "gateway_ref")
    private String gatewayRef;

    @Column(name = "processed_at", nullable = false)
    private LocalDateTime processedAt;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type", length = 50)
    private TransactionType transactionType;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: frontend/src/types/transaction.ts
import type { BaseEntity, BaseFilterRequest } from "./common";
import type {
  CurrencyCode,
  PaymentMethod,
  TransactionStatus,
  TransactionType,
} from "./enum";
import type { BookingListResponse } from "./booking";

export interface TransactionResponse extends BaseEntity {
  booking: BookingListResponse;
  amount: number;
  currency: CurrencyCode;
  paymentMethod: PaymentMethod;
  status: TransactionStatus;
  gatewayRef?: string;
  processedAt?: string;
  transactionType: TransactionType;
}

export interface TransactionRequest {
  bookingId: string;
  amount: number;
  currency: CurrencyCode;
  paymentMethod: PaymentMethod;
  gatewayRef?: string;
}

export interface TransactionFilterRequest extends BaseFilterRequest {
  bookingId?: string;
  currency?: CurrencyCode;
  paymentMethod?: PaymentMethod;
  status?: TransactionStatus;
  transactionType?: TransactionType;
  gatewayRef?: string;
  processedFrom?: string;
  processedTo?: string;
}
